### ISSUE TITLE
window['require'].on issues in esri-loader from Vuejs component

### ISSUE THREAD

**Issue by spevilgenius** [Esri/esri-loader#148]
### Expected behavior

Expect the map to be displayed in a Vuejs component

### Actual behavior

When running locally in npm the code runs fine. When the code is uploaded and ran in a production environment the error "TypeError: window.require.on is not a function" is displayed.

This error occurs on every browser and comes from the line:
var errorHandler = window['require'].on('error', reject);
It happens no matter how I load 'esri-loader' and I suspect any line the uses window['require'] will error as well.

### Steps to reproduce the behavior
I can't use typescript and prefer to use vuejs and regular javascript as 90% of my app is working fine. I am using webpack but also tried using global from unpkg with the same result
I must caveat this as I know what part of the problem is but not sure if I can host this code anywhere for you to get to it. As I said, it works in all browsers when testing locally, but in the environment it is running in production this error occurs. This leads me to believe there is a blocking element or something else that I am missing. I am just not sure why window['require'] works fine locally and not in production

I can only ask really is this functionality needed and if so is there another way to write that code so it doesn't use window['require']?

**Comment by tomwayson** [comment]
Can you provide the versions of esri-loader, JSAPI, and Vue.js that you are using?

I think it's interesting that this is the same error message as in #142. 

> I can only ask really is this functionality needed and if so is there another way to write that code so it doesn't use window['require']?

`window['require']` is needed, the error you're seeing is a symptom of a larger problem. See https://github.com/Esri/esri-loader/issues/142#issuecomment-443806477 for some troubleshooting starting points.

> I can't use typescript and prefer to use vuejs and regular javascript as 90% of my app is working fine. 

You should **not** need to use TS.

> When the code is uploaded and ran in a production environment the error.

Do you see the same error if you run in "production" locally?

> I must caveat this as I know what part of the problem is but not sure if I can host this code anywhere for you to get to it. 

See https://github.com/Esri/esri-loader/issues/144#issuecomment-450665192 for a suggestion of how to create a test case for Vue.js. Since your issue only happens in "production" it might not show up in codesandbox, but you should be able to pull that code locally and run a "production" build and verify if you see the error.

**Comment by spevilgenius** [comment]
esri-loader is the latest one 2.6 I think. JSAPI 4.9 or 4.10 both have same result. Vue version is 2>. 
When I say locally, I am referring to npm run dev using webpack dev server. It works fine there.

I did try lazy loading and tried using the global from unpkg but always the same error. I am using es6-promise/auto at the top of main.js as my promise implementation so not sure if it doesn't like it. So I tried to use the bluebird promise and that didn't work well at all.

I find it interesting that the code is written window['require'].on as when I test this using typeof command, it returns that it is indeed a function. But the error is window.require.on and testing this with typeof returns undefined. Not sure that helps any, but when I can I wanted to see if just doing a timer to keep testing that to see if it ever returns as a function. Not sure what else to try!!

**Comment by tomwayson** [comment]
> as when I test this using typeof command, it returns that it is indeed a function. But the error is window.require.on and testing this with typeof returns undefined

That is interesting. How/when are you executing those tests? In the developer tools console? At a breakpoint, or after the code executes and you see an error?

Can you try an [older version of the JSAPI](https://github.com/Esri/esri-loader#from-a-specific-version)?

Can you try running a production build locally (i.e. don't use webpack dev server)?

Are you using a manually added `<script>` tag w/ `data-esri-loader`?

Are you calling `loadModules()` more than once?

Can you please provide a snippet of code that shows how/when (i.e. where in vue component) you invoke `loadModules()`?

**Comment by spevilgenius** [comment]
Here is the script portion of the vue component:

<script>
/* eslint-disable */
import { loadModules } from 'esri-loader'
var instance = null
var map
export default {
  name: 'shop',
  data: function () {
    return {
      items: [],
      site: '',
      list: '',
      action: '',
      dlgid: '',
      dlgbody: '',
      title: '',
      map: null,
      latitude: null,
      longitude: null,
      geometry: null,
      feature: null,
      listsexist: false,
      ready: false,
      isdev: true,
      message: null,
      count: 0,
      total: 0
    }
  },
  created: function () {
    instance = this
  },
  methods: {
    drawMap: function () {
      var test = typeof window.require.on
      console.log('typeof test: ' + test)
      loadModules([
        'esri/Map',
        'esri/views/MapView'
      ], {
          // use a specific version instead of latest 4.x
          url: 'https://js.arcgis.com/4.10/'
        }).then(([Map, MapView]) => {
          console.log('Modules loaded!')
          map = new Map({
            basemap: 'streets'
          })

          const view = new MapView({
            container: 'viewDiv',
            map,
            zoom: 3,
            center: [0, 0] // longitude, latitude
          })
        })
    },
    setSearchpopup: function () {
      this.$refs['viewmodal'].show()
    },
    hidepopup: function () {
      this.$refs['missionmodal'].hide()
    },
    addStore: function () {
      alert('Add Store for point at ' + this.latitude + ', ' + this.longitude)
    }
  },
  mounted: function () {
    console.log('component mounted')
    instance.drawMap()
  }
}
</script>

I know I am using loadModules() only once. I did try to use the script tag with the loaded attribute and still same result. I was going to try to do a setInterval test to see if there is a timing issue because it will "almost" work if I put a breakpoint on .then and wait before continuing.

**Comment by andygup** [comment]
Assuming you aren't doing server-side rendering, whatabout trying something like this?

```
mounted: function () {
  this.$nextTick(function () {
    instance.drawMap()
  })
}
```

**Comment by spevilgenius** [comment]
I thought about that in some other components but not this one. I did try it but got the same result unfortunately.

**Comment by andygup** [comment]
What about not setting `this` from within `created`? I wonder if that's locking the scope to the component state snapshot during the created event.

**Comment by spevilgenius** [comment]
I did try to switch to just using `this` in lieu of setting a variable like instance but the result is the same. I usually have some async calls for loading data which is why I use an instance variable in that way. Here I just want to see a map. I have looked over several examples and am curious to understand if a dojo config is needed or not. I don't have one, but was under the impression that I didn't need one in this use case. I will continue to try other things as well.

**Comment by tomwayson** [comment]
@spevilgenius did you try using JSAPI 4.7 yet? 

The line you are seeing the error on is actually **not** the same line as in #142 [as I assumed above](https://github.com/Esri/esri-loader/issues/148#issuecomment-452130153). It actually is the same line as in #147: 

https://github.com/Esri/esri-loader/blob/cfcc200989dd590f3272f0166bcb28899a12df26/src/esri-loader.ts#L146

So it would be very helpful to know if you see this problem w/ <= v4.7 of the ArcGIS API.

**Comment by tomwayson** [comment]
closing due to inactivity

**Comment by alanscandrett** [comment]
### **tl;dr:**
- Dojo.js has a conditional that sets window.required
- This conditional will leave window.required as undefined if window.define exists.
- This can be caused by dojo.js loading at the same time as another library which sets window.define.

**...**
I had the same issue on a clients React-Typescript project, I had an issue similar to #148 and #144, specifically that. After moving through the numerous stages of grief whilst scouring the internet until the early morning I arrived at the following conclusion:

It turns out that within Dojo.js there is a small conditional statement that assigns a value to window.require, but only if there is not an existing value on window.define. Below is the code in question (minified but beautified):
![image](https://user-images.githubusercontent.com/23227595/74855559-96473f80-538c-11ea-84a6-a762b530c689.png)

This is caused, somehow, by some define statements used elsewhere in the clients code around the same time as the dojo file is running. I am on the path of finding and eliminating this in my project with extreme prejudice. 

So for anyone else with the same issue it's unlikely to be an issue with the libraries themselves, but rather a conflict in your mixture of different libs and their dependencies.

**Comment by alanscandrett** [comment]
tl;dr:
dojo.js has a conditional that sets window.required
This conditional will leave window.required as undefined if window.define exists.
This can be caused by dojo.js loading after another library has set window.define
Code:
window.define || (window.define = P, window.require = t)

Long Version
![image](https://user-images.githubusercontent.com/23227595/74898709-6593f400-53e6-11ea-9d57-812d92da453b.png)

I had the same issue with a React-Typescript project, I had an issue similar to #148 and #144, specifically that. I arrived at the following conclusion:

It turns out that within dojo.js there is a small conditional statement that assigns a value to window.require, but only if there is not an existing value on window.define.
Below is the code in question:

![image](https://user-images.githubusercontent.com/23227595/74898680-50b76080-53e6-11ea-8dd4-27b5fdfc0759.png)

This is caused, somehow, by some define statements used elsewhere in the clients code (https://www.npmjs.com/package/@microsoft/sp-build-web I believe) beforethe dojo file has run.
I don't see a resolution to this other than a hacky setting of window.define to undefined and back again after dojo.js loads (Just before I call loadModules), and it seems that the code will set window.define back again itself anyway. 

So for anyone else with the same issue it's unlikely to be an issue with the libraries themselves, but rather a conflict between your mixture of different libs/dependencies, and dojo during its loading period.

Though not fully understanding Dojo I could be quite wrong, if so please correct me! Is there a way around this perhaps for those using libraires like this?

**Comment by tomwayson** [comment]
@alanscandrett 

Thank you for the detailed information.

If you've isolated the issue to a specific library that is causing a problem by setting `window.define` before Dojo is loaded, you might try:
1. does the library have a "no conflict" mode?
1. does the library have an AMD distribution that can be loaded by Dojo's loader (see [configuring dojo](https://github.com/Esri/esri-loader#configuring-dojo))

Alternatively, I agree that this is hacky:

>setting of window.define to undefined and back again after dojo.js loads (Just before I call loadModules), and it seems that the code will set window.define back again itself anyway.

but if it works, I'd ship it.
